libraries/AC_AutoTune/AC_AutoTune_Heli.h

```diff
@@ -0,0 +1,285 @@
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/*
+  support for autotune of helicopters
+ */
+
+#pragma once
+
+#include "AC_AutoTune.h"
+
+class AC_AutoTune_Heli : public AC_AutoTune
+{
+public:
+    // constructor
+    AC_AutoTune_Heli();
+
+    // save gained, called on disarm
+    void save_tuning_gains() override;
+
+    // var_info for holding Parameter information
+    static const struct AP_Param::GroupInfo var_info[];
+
+protected:
+
+    //
+    // methods to load and save gains
+    //
+
+    // backup original gains and prepare for start of tuning
+    void backup_gains_and_initialise() override;
+
+    // switch to use original gains
+    void load_orig_gains() override;
+
+    // switch to gains found by last successful autotune
+    void load_tuned_gains() override;
+
+    // load gains used between tests. called during testing mode's update-gains step to set gains ahead of return-to-level step
+    void load_intra_test_gains() override;
+
+    // load test gains
+    void load_test_gains() override;
+
+    // reset the test vaariables for heli
+    void reset_vehicle_test_variables() override;
+
+    // reset the update gain variables for heli
+    void reset_update_gain_variables() override;
+
+    // initializes test
+    void test_init() override;
+
+    // runs test
+    void test_run(AxisType test_axis, const float dir_sign) override;
+
+    // update gains for the rate p up tune type
+    void updating_rate_p_up_all(AxisType test_axis) override;
+
+    // update gains for the rate d up tune type
+    void updating_rate_d_up_all(AxisType test_axis) override;
+
+    // update gains for the rate d down tune type
+    void updating_rate_d_down_all(AxisType test_axis) override {};
+
+    // update gains for the rate ff up tune type
+    void updating_rate_ff_up_all(AxisType test_axis) override;
+
+    // update gains for the angle p up tune type
+    void updating_angle_p_up_all(AxisType test_axis) override;
+
+    // update gains for the angle p down tune type
+    void updating_angle_p_down_all(AxisType test_axis) override {};
+
+    // update gains for the max gain tune type
+    void updating_max_gains_all(AxisType test_axis) override;
+
+    // set gains post tune for the tune type
+    void set_gains_post_tune(AxisType test_axis) override;
+
+    // reverse direction for twitch test
+    bool twitch_reverse_direction() override { return positive_direction; }
```

Comment:
This works but is a bit non-standard.  A more AP-ish way to do this would be to have this method return "true" for multicopters and "false" for tradheli's and then change this code in AC_AutoTune::control_attitude():

```
        // reverse direction for multicopter twitch test
        positive_direction = twitch_reverse_direction();
```

.. to be ..
```

        // reverse direction for multicopter twitch test
        if (twitch_reverse_direction()) {
            positive_direction = !positive_direction;
        }
```

I'm not suggesting we should change this now though but maybe someday..